torch/_inductor/codegen/cpp.py

```diff
@@ -406,11 +406,9 @@ def _print_Round(self, expr):
     def _print_RoundDecimal(self, expr):
         assert len(expr.args) == 2
         number, ndigits = expr.args
+        cpp_type = INDEX_TYPE if isinstance(number, sympy.Integer) else "double"
         assert isinstance(ndigits, sympy.Integer)
-        pre, post = f"1e{abs(ndigits)}", f"1e-{abs(ndigits)}"
-        if ndigits < 0:
-            pre, post = post, pre
-        return f"static_cast<double>(std::nearbyint({pre} * {self.paren(self._print(number))}) * {post})"
+        return f"static_cast<{cpp_type}>(std::nearbyint(1e{ndigits} * {self.paren(self._print(number))}) * 1e{-ndigits})"
```

Comment:
For the integer case, can we avoid the round trip through float? For `ndigits >= 0` this is a no-op, and for negative ndigits we could write a small helper function that does the rounding (It's likely too complex to print here).

Reply:
> for negative ndigits we could write a small helper function that does the rounding (It's likely too complex to print here).

Can't we just do

```python
c = 10 ** -ndigits
f"{self.paren(self._print(number))} / {c} * {c};"
```

Since for integer inputs there is no rounding, integer division should be sufficient for negative `ndigits`.

Reply:
Ok, I'm wrong. We do have rounding even for integers. Not sure how you want to do this without going over to float. Ultimately, we need to apply the "round to even" scheme somehow.

Reply:
> It'd probably also be okay to error on negative `ndigits` as it's still better than we do currently.

Why error? The trip to float is not perfect, but certainly better than an error?

Reply:
On triton we're casting to a 32-bit float which only has 23 bits of significand, so anything above 8.5 million cannot be faithfully represented. This could easily lead to off by one errors, and in indexing code that might mean overflowing a buffer, or leaving uninitialized memory.